src/pkm/PK3RR.ts

```diff
@@ -0,0 +1,371 @@
+// This file was generated by make generate
+import {
+  AbilityFromString,
+  Ball,
+  ItemGen3FromString,
+  ItemGen3ToString,
+  Languages,
+  NatureToString,
+} from 'pokemon-resources'
+import { NationalDex, PokemonData } from 'pokemon-species-data'
+import * as conversion from '../conversion'
+import * as byteLogic from '../util/byteLogic'
+import { 
+  shuffleBlocksGen3,
+  decryptByteArrayGen3 
+} from '../util/encryption'
+import { genderFromPID } from '../util/genderCalc'
+import { AllPKMFields } from '../util/pkmInterface'
+
+import { getLevelGen3Onward, getStats } from '../util/statCalc'
+import * as stringLogic from '../util/stringConversion'
+import * as types from '../util/types'
+import {
+  adjustMovePPBetweenFormats,
+  generatePersonalityValuePreservingAttributes,
+} from '../util/util'
+
+export class PK3RR {
+  format: 'PK3RR' = 'PK3RR'
+  personalityValue: number
+  trainerID: number
+  secretID: number
+  languageIndex: number
+  markings: types.MarkingsFourShapes
+  dexNum: number
+  heldItemIndex: number
+  exp: number
+  movePPUps: number[]
+  trainerFriendship: number
+  moves: number[]
+  movePP: number[]
+  evs: types.Stats
+  pokerusByte: number
+  metLocationIndex: number
+  metLevel: number
+  gameOfOrigin: number
+  ball: number
+  ivs: types.Stats
+  isEgg: boolean
+  isNicknamed: boolean
+  currentHP: number = 0;
+  nickname: string
+  trainerName: string
+  trainerGender: boolean
+
+  constructor(arg: ArrayBuffer | AllPKMFields, encrypted?: boolean) {
+    if (arg instanceof ArrayBuffer) {
+      let buffer = arg
+      const dataView = new DataView(buffer)
+      
+      // https://github.com/Skeli789/Complete-Fire-Red-Upgrade/blob/master/include/new/pokemon_storage_system.h
+
+      // Personality 0:4
+      this.personalityValue = dataView.getUint32(0x0, true)
+
+      // OTID 4:8
+      this.trainerID = dataView.getUint16(0x4, true)
+      this.secretID = dataView.getUint16(0x6, true)
+
+      // Nickname 8:18
+      this.nickname = stringLogic.readGen3StringFromBytes(dataView, 0x8, 10)
+
+      // Language 18
+      this.languageIndex = dataView.getUint8(0x12)
+
+      // Sanity 19
+      // this.sanity = dataView.getUint8(0x13)
+
+      // OT Name 20:27
+      this.trainerName = stringLogic.readGen3StringFromBytes(dataView, 0x14, 7)
+
+      // Markings 27
+      this.markings = types.markingsFourShapesFromBytes(dataView, 0x1b)
+      
+      // Species 28:30 !!!
+      this.dexNum = conversion.fromGen3PokemonIndex(dataView.getUint16(0x1c, true))
```

Comment:
My understanding is that Radical Red supports the national dex, in which case this will need to take that into account. Do you know if RR still uses the same species indexes for gen 1-3 mons?

Reply:
Yes its the national dex (I think). However there are unique Sevii-forms to some pokemon. These will probably need to be excluded. Attached is the full list of pokemon:

[Species.txt](https://github.com/user-attachments/files/17604865/Species.txt)